File: sdks/csharp/DownloadRequest.cs
using System.Diagnostics;
using System.Text.Json;

namespace SourisDW;

/// <summary>
/// Download request with fluent API.
/// </summary>
public class DownloadRequest
{
    private readonly SourisDownloader _dw;
    private readonly string _url;
    private readonly string? _mediaType;
    private string? _format;
    private string? _quality;
    private string? _output;
    private int? _parallel;
    private bool? _embedMetadata;
    private bool? _embedThumbnail;
    private bool? _embedSubtitles;
    private Action<JsonElement>? _onProgress;
    private Action<JsonElement>? _onComplete;
    private Action<string>? _onError;

    internal DownloadRequest(SourisDownloader dw, string url, string? mediaType = null)
    {
        _dw = dw;
        _url = url;
        _mediaType = mediaType;
    }

    public DownloadRequest Format(string fmt) { _format = fmt; return this; }
    public DownloadRequest Quality(string q) { _quality = q; return this; }
    public DownloadRequest Output(string path) { _output = path; return this; }
    public DownloadRequest Parallel(int n) { _parallel = n; return this; }
    public DownloadRequest EmbedMetadata(bool enabled) { _embedMetadata = enabled; return this; }
    public DownloadRequest EmbedThumbnail(bool enabled) { _embedThumbnail = enabled; return this; }
    public DownloadRequest EmbedSubtitles(bool enabled) { _embedSubtitles = enabled; return this; }
    public DownloadRequest OnProgress(Action<JsonElement> callback) { _onProgress = callback; return this; }
    public DownloadRequest OnComplete(Action<JsonElement> callback) { _onComplete = callback; return this; }
    public DownloadRequest OnError(Action<string> callback) { _onError = callback; return this; }

    public async Task<JsonElement> RunAsync()
    {
        var config = _dw.Config;
        var args = new List<string> { "download", _url, "--json" };

        var fmt = _format ?? config.Format;
        if (fmt != null) { args.Add("--format"); args.Add(fmt); }

        var q = _quality ?? config.Quality;
        if (q != null) { args.Add("--quality"); args.Add(q); }

        var outDir = _output ?? config.Output;
        if (outDir != null) { args.Add("--output"); args.Add(outDir); }

        var p = _parallel ?? config.Parallel;
        if (p > 0) { args.Add("--parallel"); args.Add(p.ToString()); }

        var meta = _embedMetadata ?? config.EmbedMetadata;
        if (meta) args.Add("--embed-metadata");

        var thumb = _embedThumbnail ?? config.EmbedThumbnail;
        if (thumb) args.Add("--embed-thumbnail");

        var subs = _embedSubtitles ?? config.EmbedSubtitles;
        if (subs) args.Add("--embed-subtitles");

        if (_mediaType == "audio") args.Add("--audio-only");
        if (_mediaType == "video") args.Add("--video-only");

        var psi = new ProcessStartInfo
        {
            FileName = "souris-dw",
            Arguments = string.Join(" ", args),
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            CreateNoWindow = true,
        };

        using var process = Process.Start(psi);
        if (process == null) throw new Exception("Failed to start process");

        var result = new JsonElement();

        while (await process.StandardOutput.ReadLineAsync() is { } line)
        {
            line = line.Trim();
            if (string.IsNullOrEmpty(line)) continue;

            try
            {
                var doc = JsonDocument.Parse(line);
                var root = doc.RootElement;
                var type = root.GetProperty("type").GetString();

                switch (type)
                {
                    case "progress":
                        _onProgress?.Invoke(root);
                        break;
                    case "complete":
                        result = root;
                        _onComplete?.Invoke(root);
                        break;
                    case "error":
                        var msg = root.GetProperty("message").GetString() ?? "Unknown error";
                        _onError?.Invoke(msg);
                        throw new Exception($"Download error: {msg}");
                }
            }
            catch (JsonException)
            {
                // Skip invalid JSON
            }
        }

        await process.WaitForExitAsync();

        if (process.ExitCode != 0 && result.ValueKind == JsonValueKind.Undefined)
        {
            var error = await process.StandardError.ReadToEndAsync();
            throw new Exception($"Download failed: {error}");
        }

        return result;
    }
}

File: sdks/java/src/main/java/io/souris/sourisdw/DownloadRequest.java
package io.souris.sourisdw;

import java.io.*;
import java.util.function.Consumer;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

/**
 * Download request with fluent API.
 */
public class DownloadRequest {
    private final SourisDW dw;
    private final String url;
    private final String mediaType;
    private String format;
    private String quality;
    private String output;
    private Integer parallel;
    private Boolean embedMetadata;
    private Boolean embedThumbnail;
    private Boolean embedSubtitles;
    private Consumer<JsonObject> onProgress;
    private Consumer<JsonObject> onComplete;
    private Consumer<String> onError;

    DownloadRequest(SourisDW dw, String url, String mediaType) {
        this.dw = dw;
        this.url = url;
        this.mediaType = mediaType;
    }

    public DownloadRequest format(String fmt) {
        this.format = fmt;
        return this;
    }

    public DownloadRequest quality(String q) {
        this.quality = q;
        return this;
    }

    public DownloadRequest output(String path) {
        this.output = path;
        return this;
    }

    public DownloadRequest parallel(int n) {
        this.parallel = n;
        return this;
    }

    public DownloadRequest embedMetadata(boolean enabled) {
        this.embedMetadata = enabled;
        return this;
    }

    public DownloadRequest embedThumbnail(boolean enabled) {
        this.embedThumbnail = enabled;
        return this;
    }

    public DownloadRequest embedSubtitles(boolean enabled) {
        this.embedSubtitles = enabled;
        return this;
    }

    public DownloadRequest onProgress(Consumer<JsonObject> callback) {
        this.onProgress = callback;
        return this;
    }

    public DownloadRequest onComplete(Consumer<JsonObject> callback) {
        this.onComplete = callback;
        return this;
    }

    public DownloadRequest onError(Consumer<String> callback) {
        this.onError = callback;
        return this;
    }

    public JsonObject run() throws Exception {
        Config config = dw.getConfig();

        ProcessBuilder pb = new ProcessBuilder();
        pb.command().add("souris-dw");
        pb.command().add("download");
        pb.command().add(url);
        pb.command().add("--json");

        String fmt = format != null ? format : config.getFormat();
        if (fmt != null) {
            pb.command().add("--format");
            pb.command().add(fmt);
        }

        String q = quality != null ? quality : config.getQuality();
        if (q != null) {
            pb.command().add("--quality");
            pb.command().add(q);
        }

        String out = output != null ? output : config.getOutput();
        if (out != null) {
            pb.command().add("--output");
            pb.command().add(out);
        }

        int p = parallel != null ? parallel : config.getParallel();
        if (p > 0) {
            pb.command().add("--parallel");
            pb.command().add(String.valueOf(p));
        }

        boolean meta = embedMetadata != null ? embedMetadata : config.isEmbedMetadata();
        if (meta) pb.command().add("--embed-metadata");

        boolean thumb = embedThumbnail != null ? embedThumbnail : config.isEmbedThumbnail();
        if (thumb) pb.command().add("--embed-thumbnail");

        boolean subs = embedSubtitles != null ? embedSubtitles : config.isEmbedSubtitles();
        if (subs) pb.command().add("--embed-subtitles");

        if ("audio".equals(mediaType)) pb.command().add("--audio-only");
        if ("video".equals(mediaType)) pb.command().add("--video-only");

        Process proc = pb.start();
        JsonObject result = new JsonObject();

        BufferedReader reader = new BufferedReader(new InputStreamReader(proc.getInputStream()));
        String line;
        while ((line = reader.readLine()) != null) {
            line = line.trim();
            if (line.isEmpty()) continue;

            try {
                JsonObject event = JsonParser.parseString(line).getAsJsonObject();
                String type = event.has("type") ? event.get("type").getAsString() : "";

                switch (type) {
                    case "progress":
                        if (onProgress != null) onProgress.accept(event);
                        break;
                    case "complete":
                        result = event;
                        if (onComplete != null) onComplete.accept(event);
                        break;
                    case "error":
                        String msg = event.has("message") ? event.get("message").getAsString() : "Unknown error";
                        if (onError != null) onError.accept(msg);
                        throw new Exception("Download error: " + msg);
                }
            } catch (com.google.gson.JsonSyntaxException e) {
                // Skip invalid JSON
            }
        }

        int exitCode = proc.waitFor();
        if (exitCode != 0 && !result.has("type")) {
            String stderr = new String(proc.getErrorStream().readAllBytes());
            throw new Exception("Download failed: " + stderr);
        }

        return result;
    }
}
